narwhal/primary/src/core.rs

```diff
@@ -608,19 +610,30 @@ impl Core {
             .certificates_processed
             .with_label_values(&[certificate_source])
             .inc();
+
         // Append the certificate to the aggregator of the
         // corresponding round.
-        self.append_certificate_in_aggregator(certificate.clone())
-            .await?;
+        let digest = certificate.digest();
+        if let Err(e) = self
+            .append_certificate_in_aggregator(certificate.clone())
+            .await
+        {
+            warn!(
+                "Failed to aggregate certificate {} for header: {}",
+                digest, e
+            );
+            return Err(DagError::ShuttingDown);
+        }
```

Comment:
If you don't want to repeat the `DagError::ShuttingDown` you can instead use the `tap_err` method to use the returned `Result` and do:

```suggestion
        self.append_certificate_in_aggregator(certificate.clone())
            .await
            .tap_err(|err| {
                warn!(
                    "Failed to aggregate certificate {} for header: {}",
                    digest, err
                )
            })?;
```

Reply:
Good point, this is a good alternative. But I do want to make sure only `DagError::ShuttingDown` is returned from this point. Although this is indeed the only possible error from `append_certificate_in_aggregator()`, it is not obvious.